python/desc/imsim/flats.py

```diff
@@ -0,0 +1,90 @@
+"""
+import galsim
+import desc.imsim
+
+__all__ = ['make_flat']
+
+
+def make_flat(gs_det, counts_per_iter, niter, rng, buf=2, logger=None,
+              wcs=None):
+    """
+    Create a flat by successively adding lower level flat sky images.
+    This is based on
+    https://github.com/GalSim-developers/GalSim/blob/releases/2.0/devel/lsst/treering_flat.py
+
+    Full LSST CCDs are assembled one amp at a time to limit the memory
+    used by the galsim.SiliconSensor model.
+
+    Parameters
+    ----------
+    gs_det: GalSimDetector
+        The detector in the LSST focalplane to use.  This object
+        contains the CCD pixel geometry, WCS, and treering info.
+    counts_per_iter: int
+        Roughly the number of electrons/pixel to add at each iteration.
+    niter: int
+        Number of iterations. Final flat will have niter*counts_per_iter
+        electrons/pixel.
+    rng: galsim.BaseDeviate
+        Random number generator.
+    buf: int [2]
+        Pixel buffer around each to account for charge redistribution
+        across pixel boundaries.
+    logger: logging.Logger [None]
+        Logging object. If None, then a default logger will be used.
+    wcs: galsim.fitswcs.GSFitsWCS [None]
+        Alternative WCS object. If None, then gs_det.wcs will be used.
+
+    Returns
+    -------
+    galsim.ImageF
+    """
+    if logger is None:
+        logger = desc.imsim.get_logger('DEBUG', name='make_flat')
+    if wcs is None:
+        wcs = gs_det.wcs
+    ncol = gs_det.xMaxPix - gs_det.xMinPix + 1
+    nrow = gs_det.yMaxPix - gs_det.yMinPix + 1
+    flat = galsim.ImageF(ncol, nrow, wcs=wcs)
+    sensor = galsim.SiliconSensor(rng=rng,
+                                  treering_center=gs_det.tree_rings.center,
+                                  treering_func=gs_det.tree_rings.func,
+                                  transpose=True)
+
+    # Create a noise-free base image to add at each iteration.
+    base_image = galsim.ImageF(ncol, nrow, wcs=wcs)
+    base_image.wcs.makeSkyImage(base_image, sky_level=1.)
+    mean_pixel_area = base_image.array.mean()
+    sky_level_per_iter = counts_per_iter/mean_pixel_area
+    base_image *= sky_level_per_iter
+
+    noise = galsim.PoissonNoise(rng)
+
+    # Build up the full CCD by amplifier segment to limit the memory
+    # used by the silicon model.
+    nx, ny = 2, 8
+    dx = ncol//nx
+    dy = nrow//ny
+    for i in range(nx):
+        xmin = i*dx + 1
+        xmax = (i + 1)*dx
+        for j in range(ny):
+            logger.info("amp: %d, %d", i, j)
+            ymin = j*dy + 1
+            ymax = (j + 1)*dy
+            temp_image = galsim.ImageF(ncol, nrow, wcs=wcs)
+            bounds = (galsim.BoundsI(xmin-buf, xmax+buf, ymin-buf, ymax+buf)
+                      & temp_image.bounds)
+            temp_amp = temp_image[bounds]
+            for it in range(niter):
+                logger.debug("iter %d", it)
+                temp = sensor.calculate_pixel_areas(temp_amp)
```

Comment:
@rmjarvis From the cProfile output, this pixel area calculation is the main hot spot.  I assume there's no reason not to move this outside of the `for it` loop?

Reply:
Moving it outside the for loop, reduces the execution time for the default parameters to ~5 minutes.

Reply:
No.  You can't move it outside the for loop.  That's the thing that accounts for the brighter-fatter as the flux is building up.  Moving it out of the for loop would be equivalent to setting niter=1.  (And readjusting counts_per_iter.)

But that's good that this is the hot spot.  I think that's what we expect, since it's doing a fair bit of calculation, figuring out the area of the pixel polynomials.  And I already spent some time optimizing the C++ code that does this.  There might be 10% or so more improvements possible, but I don't think there are large factors of speed up left in that code.

Reply:
I went back and looked at our original conversation in slack when we first discussed this.. Mike said 

"The script to do this is on the silicon_speedup branch now.  (I just pushed it -- I think it's all working properly now.)  It makes 10 flats.  As written, it takes about 15 minutes on my laptop (for all 10).   It has tree rings turned on.  As well as a non-trivial WCS.  Mostly to make sure these work properly.  So you can either turn them off or take differences."

Then, Jim ran  a sensor in about  ~15 minutes and we decided most of this was due to the LSST WCS.  

For dome flats at sky level where we are below the level where BF kicks in, we should be able to get away with not recalculating those shapes right?  For running in that model can we just use niter=1?

Reply:
OK, so now I realize the difference in what I was remembering for timing. Those were for actual sky levels where we were below BF and this is what is being used in imSim for the sky.

What level do we need here? 10K?

Reply:
80K is the default in Jim's script, which is pretty typical for real dome flats.  He does 20 iterations, which seems like a good choice to me.  (It matches what I did in my version of this script.)

I don't see any reason to do low-intensity dome flats for imsim, but if you wanted to do something that tries to mimic the behavior of sky flats, then yes setting niter=1 would do it.

Reply:
I thought we needed both high intensity for BF (80K is good for that) and low intensity for treerings, flat fielding etc. Do you mean since there is no vignetting it isn't needed here?

Reply:
I mean low-intensity dome flats aren't normally a thing.  We're just doing that for PhoSim, since it's exorbitant to run flats.  The high intensity ones do everything we need for imsim.

Reply:
Here's the output of running with cProfile for a 4 iteration run with the pixel area calculation moved back into the loop:
```
In [3]: stats.strip_dirs().sort_stats('tottime').print_stats(10)
Sat Aug 11 20:06:37 2018    make_flats_cProfile.txt

         42367001 function calls (42251154 primitive calls) in 614.059 seconds

   Ordered by: internal time
   List reduced from 10675 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       64  340.360    5.318  340.435    5.319 sensor.py:273(calculate_pixel_areas)
      401   30.455    0.076   30.455    0.076 LsstCameraUtils.py:643(<listcomp>)
      401   29.950    0.075   29.950    0.075 LsstCameraUtils.py:647(<listcomp>)
      401   29.363    0.073   29.363    0.073 CameraUtils.py:1031(<listcomp>)
      401   26.412    0.066   26.412    0.066 CameraUtils.py:1029(<listcomp>)
      402   17.127    0.043  253.825    0.631 galSimCameraWrapper.py:616(_raDecFromPixelCoords)
      403   13.241    0.033  197.801    0.491 LsstCameraUtils.py:571(pupilCoordsFromPixelCoordsLSST)
    24665   12.507    0.001   12.508    0.001 {built-in method numpy.core.multiarray.array}
      402   11.135    0.028   72.932    0.181 CameraUtils.py:1001(pupilCoordsFromFocalPlaneCoords)
      402   11.065    0.028   11.065    0.028 {palpy.ampqkVector}
```
All those list comprehensions in the sims code are the expensive part of the sky background calculation, which has only one iteration  Those sims calculations are done by the wcs implementation and only once. As a result, the 20 iteration timing is dominated by the sensor code even if we use the initially slower LSST wcs code.

Reply:
Thanks.  I think that all makes sense.  The LSST wcs code could still be sped up a lot I think, but not really a job for today.  I'd keep the production run at around 20 iterations.  Or specifically, I think 4000 counts per iteration is pretty good to make sure the pixel areas don't change much between iterations (and therefore the quantization of that calculation isn't causing problems).